### PR TITLE
Implement Helix Support (WIP)

### LINKED ISSUE #4642: Helix keymap
I know it can be tricky, but the only reason preventing me from day to day use of zed is Helix with their beautiful system based on keys without mouse, multiple selections and other features.

Will be great to support that and maybe I will try to research possible ways to do so:
https://docs.helix-editor.com/keymap.html

### PR DESCRIPTION
**Pull request by wmstack**
Closes #4642 

- Added the ability to switch to helix normal mode, with an additional helix visual mode.
- <kbd>ctrl</kbd><kbd>h</kbd> from Insert mode goes to Helix Normal mode. <kbd> i </kbd> and <kbd> a </kbd> to go back.
- Need to find a way to perform the helix normal mode selection with <kbd> w </kbd>, <kbd>e </kbd>, <kbd> b </kbd> as a first step. Need to figure out how the mode will interoperate with the VIM mode as the new additions are in the same crate.

Release notes:
- N/A

### CODE REVIEW

**Comment by wmstack** [comment]
I am trying to understand how to implement the Helix signature motions in normal mode. I think the most helpful piece of the code is this visual motion piece, which I am putting here to study:

https://github.com/zed-industries/zed/blob/ec5d6e96bb304886f49680b54e58c9b0e7221cd6/crates/vim/src/visual.rs#L118-L197

**Comment by wmstack** [comment]
@cla-bot check

**Comment by wmstack** [comment]
Okay, I have implemented some semblance of Helix's `w`, `e` and `b`. The only one that doesn't work as intended is the `e` motion. It is supposed to go the the letter at the end of the word, with the tail at the old head's position.

**Comment by maan2003** [comment]
@wmstack I had implemented lot of helix bindings in my old PR #13768

**Comment by ConradIrwin** [comment]
Thanks for this! I'm struggling to decide what to do with Helix: there's clearly a lot of demand for it, but I know that there's enough to do in vim to keep me busy forever :D.

As mentioned before, I think it'd be ok to merge a WIP helix mode in such a way that it's not impacting the overall user/developer experience, and I think this PR is a reasonable approach (#17575 goes a completely different way, and I also think that is reasonable - but this seems less intrusive somehow).

Back to your questions:

One thing that causes problems in vim (and probably will in helix too) is that a collapsed selection in zed is represented as the position before the character that the block cursor is under; whereas an expanded selection spans from before the first character to after the last.

In vim you can only be in normal mode when selections are collapsed, but in helix, normal mode may have expanded or collapsed selections. https://github.com/zed-industries/zed/blob/56163b1e3595b98986a2ada8d54df8b72bfc6187/crates/vim/src/visual.rs#L142-L193 is probably the logic you need to copy (conditional on whether the selections are expanded or collapsed).

If you'd like to pair on this, please book time here: https://calendly.com/conradirwin/pairing (@maan2003 let me know if you want to join too, and I'll send you the link).

**Comment by baldwindavid** [comment]
> I'm struggling to decide what to do with Helix: there's clearly a lot of demand for it, but I know that there's enough to do in vim to keep me busy forever :D.

I use and like Helix a lot, but part of its allure is the opinionated and constrained package. Dumping some Helix concepts into Zed doesn't really make it Helix and potentially adds a lot of expectations and confusion in terms of maintenance and feature parity. I wonder if it's possible to codify the name of the feature being added for clarity and to limit expectations.

From what I can tell the desire is mostly for the object-verb grammar as opposed to verb-object. Verb-object is not even a Helix thing; it's lifted from Kakoune. At any rate, I wonder if this could be slotted into the existing vim settings with something like:

```jsonc
"vim_mode": true,
"vim": {
  "grammar": "helix (or kakoune)" // default "vim" (or vi if I'm being so pedantic :))
},
```

or

```jsonc
"vim_mode": true,
"vim": {
  "grammar": "object_verb" // default "verb_object"
},
```

If that doesn't prove to be enough then perhaps in the future a more full-blown Helix/Kak extension similar to VSCode's Dance emerges down the road.

**Comment by wmstack** [comment]
Okay word movements `w` `e` and `b` are working as expected:


https://github.com/user-attachments/assets/59fff066-16a4-477a-81f0-0b5b2d01c00e

**Comment by ConradIrwin** [comment]
Nice!

Before I merge this I'd like to talk things through with you. 

Also, I think we should:
* add a test
* remove the HelixVisual placeholder
* remove the `ctrl-h` keybinding and add something to the docs explaining how people can try this,

**Comment by wmstack** [comment]
Apologies for the delay. When I first submitted the pull request, I had upcoming university exams that required my focus. After completing those exams, I found it challenging to regain the motivation to continue with this. Thanks for your patience.

I added a test for next word start, and I discovered that it wasn't actually working as intended for the word before a new line it selected the newline character too. So I fixed that.

Also, I am aware of a problem where when moving one character right when there is a selection, a character is skipped (as can be seen in the video above). This probably happens for the reason you mentioned before. However, as far as I am aware, the word motions are all correct, but I haven't tested them extensively as their are a number of edges cases.

I have booked a meeting as I am interested to discuss the next steps. I feel like the issue should be open until a significant number of helix commands are functional.

File: assets/keymaps/vim.json [hunk 1]
```diff
@@ -321,7 +321,6 @@
     "context": "vim_mode == insert",
     "bindings": {
       // experimental keybinding to toggle helix mode
```
Comment (dunxen) [inline assets/keymaps/vim.json:323]: nit: remove this comment that was left behind :)

Comment (wmstack) [reply assets/keymaps/vim.json]: Thanks for catching that.

**Comment by wmstack** [comment]
@ConradIrwin Well I kind of don't think that the `boundary_trail` functions should be separated, as both the head and the tail move using the same logic. I have changed the functions to take only the head as input, so as to simplify its mechanics a bit. 

However, really, if each function is split into two, we would have 4 functions, and both parts would really look almost identical, they would share much of the same functionality, namely skipping newline characters, and skipping the first boundary if it lies just after the newlines. The bulk of the code, which determines where to put the new tail at, is also used to move the head forward to start testing for boundaries starting from the correct point.

**Comment by maan2003** [comment]
try borrowing some tests from helix btw

lots of tests there https://github.com/helix-editor/helix/blob/master/helix-core/src/movement.rs

**Comment by ConradIrwin** [comment]
@wmstack still a couple of formatting things to fix

**Comment by BrianTheMessiah** [comment]
https://app.opire.dev/issues/01J6HJZCK02FVD4XG900FY36EE is this bounty still available or has @ConradIrwin claimed it?

**Comment by ConradIrwin** [comment]
@BrianTheMessiah I don't think it's been completed yet.

@wmstack I am going to merge this PR, sorry for the delay, I was out last week.

For anyone wanting to try the (very nascent) helix mode out, you can currently access it by creating a key-binding that does `["vim::SwitchMode", "helix_normal"]`. A lot more work is needed before we can really expect people to use this though help is welcome :D.

**Comment by well1791** [comment]
is this available on stable release? or preview release? or just by building it locally?

**Comment by kranurag7** [comment]
> is this available on stable release? or preview release? or just by building it locally?

from the time it was merged which is today, we didn't had any pre-release or stable release of building from main would have it.

**Comment by baldwindavid** [comment]
Assuming the typical Wednesday Preview Release, it will be in there.

**Comment by nabby27** [comment]
hI @wmstack I am the creator of the bounty mentioned and seeing your great work and (although the integration is not complete) the issue has already been closed and your PR has been merged, I would be more than happy to pay you the bounty for your effort, you would only need to claim it in Opire and configure your Stripe account to receive the money.

**Comment by Siloss** [comment]
> For anyone wanting to try the (very nascent) helix mode out, you can currently access it by creating a key-binding that does `["vim::SwitchMode", "helix_normal"]`. A lot more work is needed before we can really expect people to use this though help is welcome :D.

isn't it `["vim::SwitchMode", "HelixNormal"]`? Only that option worked for me.

**Comment by wmstack** [comment]
The way I did the command was
```
"bindings": {
      "escape": "vim::HelixNormalAfter"
}
```

**Comment by ricardonunez-io** [comment]
Is this still in active development to reach a state of parity or almost-parity with Helix? 

Happy to add to the bounty if so.

**Comment by freepicheep** [comment]
I tried configuring this in my settings, but still am unable to enter Helix Normal. Can anyone help?

**Comment by forivall** [comment]
From my experimentation, this is still very work in progress, but it's promising! My config is [here](https://github.com/forivall/dotfiles/blob/osx/config/zed/keymap.json) for those who want to use it, although it's missing many of my most common motions and actions. It's good enough for a bit of light editing.

**Comment by freepicheep** [comment]
@forivall thanks! That is a great starting place. If it's useful, I immediately added this keybind for the `c` movement from Helix `"c": ["workspace::SendKeystrokes", "d i"]` under `"context": "vim_mode == helix_normal && !menu"`. 

Is there any way to make Helix normal the default instead of (vim) normal?

**Comment by neel04** [comment]
Is there any way we can manually add some of the more complex keybindings, such as `Alt/Option + o` for selecting syntactic tokens of code through tree-sitter?

**Comment by nyabinary** [comment]
What the remainder of the work needed?

**Comment by ConradIrwin** [comment]
Roughly:
* Creating a setting that enables HelixNormal instead of Normal by default. (Maybe #25067 would be enough)
* Bundling as many of Helix's bindings as we can reasonably support.
* Adding more actions to support any bindings that are deemed reasonable, but not implementatable yet.

I think a few people have said they are using helix-like keybindings, so maybe that'd be a good starting point for 2.

**Comment by nyabinary** [comment]
> Roughly:
> 
>     * Creating a setting that enables HelixNormal instead of Normal by default. (Maybe [vim: Add ability to change default mode #25067](https://github.com/zed-industries/zed/pull/25067) would be enough)
> 
>     * Bundling as many of Helix's bindings as we can reasonably support.
> 
>     * Adding more actions to support any bindings that are deemed reasonable, but not implementatable yet.
> 
> 
> I think a few people have said they are using helix-like keybindings, so maybe that'd be a good starting point for 2.

Does Helix have a crate or something we can use instead for keybindings?

**Comment by storopoli** [comment]
> Does Helix have a crate or something we can use instead for keybindings?

Yes, here: https://github.com/helix-editor/helix/blob/master/helix-term/src/commands.rs

Which is a used in https://github.com/helix-editor/helix/blob/master/helix-term/src/keymap.rs

**Comment by nyabinary** [comment]
Pretty afraid Helix support is just going to rot and get eventually removed if someone doesn't step up :P.

**Comment by theherk** [comment]
I just started testing this out in [e15d20a4](https://github.com/theherk/commons/commit/e15d20a4305578ea3c29917bbf1a6565839acf06), and it actually feels really nice. I've never been a big helix / kakoune user, but having dabbled I get the draw. The way this is implemented actually feels like the best of both worlds. Nice work.

**Comment by hisirdox** [comment]
@ricardonunez-io 
> Is this still in active development to reach a state of parity or almost-parity with Helix?
> 
> Happy to add to the bounty if so.

if enough people want to use, we can do it.
i am personaly, a heavy neovim user and could work on this

**Comment by zetashift** [comment]
> @ricardonunez-io
> 
> > Is this still in active development to reach a state of parity or almost-parity with Helix?
> > Happy to add to the bounty if so.
> 
> if enough people want to use, we can do it. i am personaly, a heavy neovim user and could work on this

Helix user here, the helix bindings in Zed are a bit rough, it doesn't have default keybindings that work well with Zeds multicursors and the keymap of `m`(match around, inside and surround) is hard to replicate. Also `f` and `t` don't seem to be working.

So it depends on what people expect from helix keybindings. Personally I'd already be pretty happy if the match thingy gets parity.

**Comment by hisirdox** [comment]
i think the user expects the same use of helix, so they can just migrate to zed
not that it makes sense, or i am right on my view of the situation

**Comment by zetashift** [comment]
> i think the user expects the same use of helix, 

I get what you're saying, but even the Vim mode in Zed has some tiny differences and helix includes a lot out of the box, e.g. jumping to a word in the visible area of the buffer. Users chose helix for different kinds of reasons.

I don't think the first thing to implement is the multicursor feature of helix (`s`) because there is no real Zed equivalent for that yet. Instead ergonomic keybindings to the current Zed cursor functionality would be a better trade-off.

Instead I'd recommend more keybindings that enable `select->act` functionality are easy wins.
